frontend/src/components/worksheets/items/SchemaItem/SchemaItem.js

```diff
@@ -0,0 +1,343 @@
+}> {
+    /** Constructor. */
+    constructor(props) {
+        super(props);
+        this.state = {
+            showSchemaDetail: false,
+            rows: [...this.props.item.field_rows],
+            editing: false,
+        };
+    }
+
+    toggleEdit = (clear, save) => () => {
+        if (!this.props.editPermission) return;
+        if (clear) {
+            this.clearChanges();
+            return;
+        }
+        this.setState({ editing: !this.state.editing });
+        if (save) {
+            this.saveSchema();
+        }
+    };
+
+    clearChanges = () => {
+        this.setState({ rows: [...this.props.item.field_rows], editing: !this.state.editing });
+    };
+
+    saveSchema = () => {
+        const { schema_name, field_rows } = this.props.item;
+        let updatedSchema = [];
+        let fromAddSchema = false;
+        let schemaBlockSourceLength = field_rows.length;
+        this.state.rows.forEach((fields) => {
+            if (!fields.field) {
+                return;
+            }
+            if (!fromAddSchema && fields.from_schema_name !== schema_name) {
+                // these rows correspond to addschema
+                fromAddSchema = true;
+                updatedSchema.push('% addschema ' + fields.from_schema_name);
+                return;
+            } else if (fromAddSchema && fields.from_schema_name !== schema_name) {
+                // These rows doesn't occupy any source lines
+                schemaBlockSourceLength -= 1;
+                return;
+            } else {
+                fromAddSchema = false;
+            }
+
+            let curRow = '% add ' + fields.field;
+            if (!fields['generated-path']) {
+                updatedSchema.push(curRow);
+                return;
+            }
+            curRow = curRow + ' ' + fields['generated-path'];
+            if (!fields['post-processing']) {
+                updatedSchema.push(curRow);
+                return;
+            }
+            curRow = curRow + ' ' + fields['post-processing'];
+            updatedSchema.push(curRow);
+        });
+        this.props.updateSchemaItem(
+            updatedSchema,
+            this.props.item.start_index,
+            schemaBlockSourceLength,
+        );
+    };
+
+    addFieldRowAfter = (idx) => (e) => {
+        if (!this.props.editPermission) return;
+        const schemaItem = this.props.item;
+        const schemaHeaders = schemaItem.header;
+        let newRow = { from_schema_name: schemaItem.schema_name };
+        schemaHeaders.forEach((header) => {
+            newRow[header] = '';
+        });
+        let curRow = this.state.rows;
+        curRow.splice(idx + 1, 0, newRow);
+        this.setState({ rows: curRow });
+    };
+
+    changeFieldValue = (idx, key) => (e) => {
+        if (!this.props.editPermission) return;
+        const { rows } = this.state;
+        let copiedRows = JSON.parse(JSON.stringify(rows));
+        copiedRows[idx][key] = e.target.value;
```

Comment:
I think it's better to avoid `JSON.parse(JSON.stringify())`. Can we do this instead?

```suggestion
        let copiedRows = {...rows, [idx]: {...copiedRows[idx], [key]: e.target.value } };
```

In other cases, if this method gets too verbose (I don't think it's too verbose in this case), we could always use a library like [immutability-helper](https://github.com/kolodny/immutability-helper) to do the task.

Reply:
Doesn't quite work, maybe because you need copiedRows to be first defined, updated to use [...rows] though

Reply:
Aren't you essentially directly mutating state here now though? (because [...rows] only makes a shallow copy -- it doesn't copy every element in rows, and  copiedRows[idx][key] ends up modifying individual elements of the array)

Reply:
I updated to use splice to first remove that reference and then add it back, so we don't directly modify the reference in state.rows, checked that modifying the copiedRows now doesn't modify the state.rows